File: python/Sort/InsertSort.py
#!/usr/bin/env python

__author__ = 'fuze'

def InsertSort(number_list):
    if not isinstance(number_list, list):
        return "Input should be a list of number"
    lenth = len(number_list)
    for i in xrange(1, lenth):
        if number_list[i]<number_list[i-1]:
            temp = number_list[i]
            number_list[i] = number_list[i-1]
            pos = i-1
            while(number_list[pos]>temp and pos>-1):
                number_list[pos+1] = number_list[pos]
                pos -= 1
            number_list[pos+1] = temp
        print str(i) + ' : ' + str(number_list)

if __name__=='__main__':
    NumberList = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]
    print "0 : " + str(NumberList)
    InsertSort(NumberList)

File: cpp/Sort/InsertSort.cpp
#include <iostream>
using std::cout;
using std::endl;

void print(int a[], int n, int p){
    cout<<p<<" : ";
    for(int i=0; i<n; i++){
        cout<<a[i]<<" ";
    }
    cout<<endl;
}

void InsertSort(int a[], int n){
    for(int i=1; i<n; i++){
        if(a[i]<a[i-1]){
            int x = a[i];
            int j = i-1;
            a[i] = a[i-1];
            while(a[j]>x){
                a[j+1] = a[j];
                j--;
            }
            a[j+1] = x;
        }
        print(a, n, i);
    }
}

int main(){
    int a[] = {5,4,3,2,1};
    print(a, 5, 0);
    InsertSort(a, 5);
}
